### PR TITLE
Try to validate read ts for all RPC requests

### PR DESCRIPTION
**Pull request by MyonKeminta**
Ref: https://github.com/pingcap/tidb/issues/57786

### CODE REVIEW

File: internal/locate/region_request.go [hunk 1]
```diff
@@ -1756,6 +1764,20 @@ func (s *RegionRequestSender) onRegionError(
 	return false, nil
 }
 
+func (s *RegionRequestSender) validateReadTS(ctx context.Context, req *tikvrpc.Request) error {
+	var readTS uint64
```
Comment (cfzjywxk) [inline internal/locate/region_request.go:1768]: Checking target store type `tikv` is enough I think.

File: internal/locate/region_request.go [hunk 2]
```diff
@@ -766,6 +770,10 @@ func (s *RegionRequestSender) SendReqCtx(
 		}
 	}
 
+	if err = s.validateReadTS(bo.GetCtx(), req); err != nil {
+		return nil, nil, 0, err
```
Comment (cfzjywxk) [inline internal/locate/region_request.go:774]: Perhaps print request logs and call stack in error logs here.

**Review by cfzjywxk — COMMENTED** [review]
Rest LGTM

File: oracle/oracles/pd.go [hunk 3]
```diff
@@ -670,22 +670,34 @@ func (o *pdOracle) getCurrentTSForValidation(ctx context.Context, opt *oracle.Op
...
-	// If we fail to get latestTS or the readTS exceeds it, get a timestamp from PD to double-check.
+func (o *pdOracle) ValidateReadTS(ctx context.Context, readTS uint64, isStaleRead bool, opt *oracle.Option) (errRet error) {
+	if readTS == math.MaxUint64 {
+		if isStaleRead {
+			return oracle.ErrLatestStaleRead{}
+		}
+		return nil
+	}
+
+	latestTSInfo, exists := o.getLastTSWithArrivalTS(opt.TxnScope)
```
[the remark this answers is not in the document]
Comment (MyonKeminta) [reply oracle/oracles/pd.go:681]: I think it's better to be removed when all code related to txn scope is being removed, to keep the code consistent.

**Review by zyguan — APPROVED** [review]

**Review by cfzjywxk — APPROVED** [review]

**Review by crazycs520 — APPROVED** [review]

**Review by cfzjywxk — APPROVED** [review]

**Review by crazycs520 — APPROVED** [review]

**Pull request merged**